bakery/src/scripts/utils.py

```diff
@@ -85,3 +87,42 @@ def parse_uri(uri):
         return None
     legacy_id, legacy_version = uri.split('@')
     return legacy_id, legacy_version
+
+
+def ensure_isoformat(timestamp):
+    """Given a timestsamp string either validate it is already ISO8601 and
+    return or attempt to convert it.
+    """
+    try:
+        # Using dateutil.parser here instead of datetime.fromisoformat as the
+        # former seems to be more robust and avoids some false negatives seen
+        # with the latter (e.g. with "2021-03-22T14:14:33.17588-05:00")
+        parser.isoparse(timestamp)
+        return timestamp
+    except ValueError:
+        # The provided timestamp needs to be converted
+        pass
+
+    # Try with a timezone offset value that we can parse
+    try:
+        return datetime.strptime(timestamp, "%Y/%m/%d %H:%M:%S %z").isoformat()
```

Comment:
I wonder if this is a bug in `dateutil`, because it's parsed differently using the method below:
```python
evil_date = '2021/04/20 13:00:00.000 GMT-5'
evil_date_notz, evil_date_tz = evil_date.rsplit(' ', 1)
parse(evil_date_notz, ignoretz=True).replace(tzinfo=gettz(evil_date_tz)).isoformat()
# => '2021-04-20T13:00:00-05:00'

Reply:
Oof. Looks like this may be a feature, not a bug. Just in case you didn't hate date and time programming before:

https://stackoverflow.com/questions/31078749/timezone-offset-sign-reversed-by-dateutil

https://github.com/dateutil/dateutil/issues/70

Reply:
"Time, why you punish me? ... You left me cryin' ... Time, I don't understand ... You ain't no friend of mine" - Hootie and the Blowfish

I had to put that song on after reading those links 🙃 . 

In any case, I think if we use `parser.parse` as a final fallback after using the approach you have below things will work as desired. I'll also run the tests in the CORGI docker image just to be sure everything passes in there as far as the system dependencies.

Reply:
[Done](https://github.com/openstax/output-producer-service/pull/335/commits/29e8857bf9a30baf2a9cf33a21e14d5ce728ff4d)